### ISSUE TITLE
[FEATURE] Support for RWX Filesystem 

### ISSUE THREAD

**Issue by ghost** [harvester/harvester#5946]
**Is your feature request related to a problem? Please describe.**
When provisioning a volume for Harvester, we find that the Persistent Volumes are not created and an error like: `failed to provision volume with StorageClass "rwx-rep2": rpc error: code = InvalidArgument desc = Shared raw block volumes are not supported”`

**Describe the solution you'd like**
We would like to use RWX volumes in filesystem mode instead of requiring RWX Block Mode volumes.

**Describe alternatives you've considered**
No alternatives.

**Additional context**
Other KubeVirt distros such as OpenShift virtualization allow any RWX volumes to be used. We would like to be able to use filesystem mode for RWX volumes on harvester as well.

**Comment by ibrokethecloud** [comment]
are you able to please confirm what is the storage class provisioner?

**Comment by Vicente-Cheng** [comment]
Hi @eshanks-pure 

I've tried creating an NFS-based RWX volume for the VM before (it's just experimental). Could you share your storage provisioner?

**Comment by ghost** [comment]
> are you able to please confirm what is the storage class provisioner?

I'm speaking specifically about the Portworx Provisioner. `pxd.portworx.com`

**Comment by Vicente-Cheng** [comment]
Hi @eshanks-pure, 
Sorry for the late reply.

I would like to clarify some points for your question.

1. Did you want to use the RWX volume from the Protworx provisioner on Harvester or downstream cluster?
Harvester cluster means the VM workload
downstream cluster means the k8s cluster on the VMs managed by Harvester

2. `failed to provision volume with StorageClass "rwx-rep2": rpc error: code = InvalidArgument desc = Shared raw block volumes are not supported”`
Could you share what log file you checked for the above logs?

3. What expectation would you like to use on the VM? A block device or A filesystem? (If you would like to use RWX volume on downstream cluster, ignore this question)

Or could you give me a more detailed example of how you want to use the RWX filesystem volume?
Thanks! cc @ibrokethecloud, @bk201

**Comment by ghost** [comment]
We specifically want to use Portworx for Harvester (The VM Workloads)
The error message you refered to in bullet 2 came from a `kubectl describe` command on the PVC.

Portworx does not provide a RWX access mode for RAW block at this time, but we do support RWX access mode for Filesystem. We would like Harvester to allow us to use a RWX Filesystem for Virtual Machine disks and not require RWX Block.

Portworx provisioner already works on the downstream cluster for RWX (filesytem) and RWO Block. We want to be able to provision VMs using Portworx RWX Filesystem mode which IS available from KubeVirt, but not available on Harvester.

**Comment by Vicente-Cheng** [comment]
> We specifically want to use Portworx for Harvester (The VM Workloads) The error message you refered to in bullet 2 came from a `kubectl describe` command on the PVC.
> 
> Portworx does not provide a RWX access mode for RAW block at this time, but we do support RWX access mode for Filesystem. We would like Harvester to allow us to use a RWX Filesystem for Virtual Machine disks and not require RWX Block.
> 
> Portworx provisioner already works on the downstream cluster for RWX (filesytem) and RWO Block. We want to be able to provision VMs using Portworx RWX Filesystem mode which IS available from KubeVirt, but not available on Harvester.

Hi @eshanks-pure, thanks for the detailed explanation.

Could you provide the PVC yaml here? (I thought you created PVC using Harvester UI, right?)
The volume creation page on the Harvester will hard-code the PVC access mode to Block. But you can still manually create the filesystem mode PVC using the kubectl and the corresponding yaml file.

BTW, if you want to use the filesystem volume for the VM, you need to manually create the `disk.img` file. We did not leverage the CDI on Harvester, so you need to manually create it.
(Reference: https://kubevirt.io/user-guide/storage/disks_and_volumes/#persistentvolumeclaim)

I use the filesystem mode volume from the NFS provisioner for some prototypes if the `Portworx` provisioner could perform the filesystem mode volume. It should work on kubevirt of Harvester.

Could you try the above suggestion? Thanks!

**Comment by ccrow42** [comment]
Hey @Vicente-Cheng 

Portworx requires the equivalent of the following packages from ubuntu installations: keyutils libnfsidmap1 nfs-common nfs-kernel-server rpcbind

These don't seem to be present on the harvester OS. Is there a way to include those in an image so I can test?

**Comment by Vicente-Cheng** [comment]
Hi @ccrow42, 
Sorry, I missed that one.

Please refer to this https://docs.harvesterhci.io/v1.3/troubleshooting/os#how-can-i-install-packages-why-are-some-paths-read-only

After that, the root filesystem is not ready-only anymore. That would help you install the corresponding packages you want.

**Comment by ccrow42** [comment]
@Vicente-Cheng Thanks Vicente, I took a look at that, but it doesn't seem to be a package manager when Elemental spins up an OS (I could just be missing the obvious). The above would work for installing simple things such as statically linked binaries, but I'm after the OSes package manager.

**Comment by Vicente-Cheng** [comment]
> @Vicente-Cheng Thanks Vicente, I took a look at that, but it doesn't seem to be a package manager when Elemental spins up an OS (I could just be missing the obvious). The above would work for installing simple things such as statically linked binaries, but I'm after the OSes package manager.

After you make the root read/write with the above steps, you can add the repo by `zypper addrepo` and use the repo to install the package you want.

For example, (please replace the repo to get better connection quality)
```
$ zypper addrepo https://free.nchc.org.tw/opensuse/distribution/leap/15.6/repo/oss/ leap-15.6
$ zypper --gpg-auto-import-keys refresh
$ zypper in keyutils libnfsidmap1 nfs-client nfs-kernel-server rpcbind
```

Or did you elaborate more about your concern with the elemental spins up the OS?

**Comment by ccrow42** [comment]
Nope, just a dumb debian/gentoo guy that didn't know about zypper. Thanks @Vicente-Cheng

**Comment by Vicente-Cheng** [comment]
No Problem!
Feel free to ping me here if you have any updates (or need to clarify something). Thanks!

**Comment by ccrow42** [comment]
Hey folks,

I have been able to successfully test Harvester 1.3.0 on our end using RWX filesystem mode after some code changes. So far it looks like changes need to be made on our end. I don't have a date for those changes, but we were able to get it working without modifying the underlying OS.

**Comment by Vicente-Cheng** [comment]
Hi @ccrow42,

I am happy to hear that.
So far, you can use the RWX filesystem mode with a 3rd-party CSI driver, right?

If anything needs to be clarified, feel free to reach out here again!

**Comment by Vicente-Cheng** [comment]
Hi @ccrow42, 
Do we still need this? I thought we should be able to use Portworx in v1.5.0.

**Comment by ccrow42** [comment]
I do not think we do. We still need to qualify RWX block support, but that seems to be the best path going forward with Harveater.
